File: 994-rotting-oranges/994-rotting-oranges.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        return channaSir(grid);
    }
private:
    
    int channaSir(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        queue<pair<int, int>> q;
        int fresh = 0;
        
        for(int i = 0; i < m; ++i){
            for(int j = 0; j < n; ++j){
                if(grid[i][j] == 2)
                    q.push({i, j});
                else if(grid[i][j] == 1)
                    ++fresh;
            }
        }
        if(q.empty()){
            if(fresh == 0) return 0;
            return -1;
        }
        
        
        int level = 0;
        while(!q.empty()){
            int levelLen = q.size();
            while(levelLen--){
                auto [i, j] = q.front();
                
                q.pop();
                
                if(i-1 >= 0 && grid[i-1][j] == 1){
                    --fresh;
                    grid[i-1][j] = 2;
                    q.push({i-1, j});
                }
                if(j-1 >= 0 && grid[i][j-1] == 1){
                    --fresh;
                    grid[i][j-1] = 2;
                    q.push({i, j-1});
                }
                if(i+1 <= m-1 && grid[i+1][j] == 1){
                    --fresh;
                    grid[i+1][j] = 2;
                    q.push({i+1, j});
                }
                if(j+1 <= n-1 && grid[i][j+1] == 1){
                    --fresh;
                    grid[i][j+1] = 2;
                    q.push({i, j+1});
                }
            }
            ++level;
        }
        if(fresh > 0) return -1;
        return level-1;
    }
};

File: 994-rotting-oranges/994-rotting-oranges.py
class Node:
    def __init__(self, x, y, time):
        self.x = x
        self.y = y
        self.time = time

class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        # Aakash Sir
        
        rows = len(grid)    # Number of rows
        cols = len(grid[0])  # Number of columns
        
        queue = deque()
        
        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == 2:
                    node = Node(i, j, 0)
                    queue.append(node)
        
        ans  = 0
        
        while len(queue):
            
            currNode = queue.popleft()
            x, y, time = currNode.x, currNode.y, currNode.time
            
            if x >= 1 and grid[x - 1][y] == 1:
                grid[x - 1][y] = 2
                node = Node(x - 1, y, time + 1)
                queue.append(node)
                ans = time + 1                
            
            if x < rows - 1 and grid[x + 1][y] == 1:
                grid[x + 1][y] = 2
                node = Node(x + 1, y, time + 1)
                queue.append(node)
                ans = time + 1
            
            if y >= 1 and grid[x][y - 1] == 1:
                grid[x][y - 1] = 2
                node = Node(x, y - 1, time + 1)
                queue.append(node)
                ans = time + 1
                
            if y < cols - 1 and grid[x][y + 1] == 1:
                grid[x][y + 1] = 2
                node = Node(x, y + 1, time + 1)
                queue.append(node)
                ans = time + 1
        
        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == 1:
                    return -1
        return ans
            